Warn when using FileRedirection operator inside if/while statements and improve new PossibleIncorrectUsageOfAssignmentOperator rule

bergmeister opened the pull request:
## PR Summary

Closes #878

- Improve existing rule about assignment operator to also not trigger for more special cases like e.g. `[NameSpace]::Method()` or `$b -match $c` on the RHS of the if expression and add optional, implicit clang suppression when extra parenthesis are used.
- Add new `PossibleIncorrectUsageOfRedirectionOperator` rule to warn when using `>` unintentionally inside `if` and `elseif` statements (e.g `if ($a > $){ }`).
- Apply the same logic to while/do-while statements.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets. Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
    - [x] Use the present tense and imperative mood when describing your changes
- [x] Summarized changes
- [x] User facing documentation needed
- [x] Change is not breaking (because the rule is new and has not shipped yet)
- [x] Make sure you've added a new test if existing tests do not effectively test the code changed
- [x] This PR is ready to merge and is not work in progress.
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

bergmeister commented on RuleDocumentation/PossibleIncorrectUsageOfComparisonOperator.md:
```diff
@@ -0,0 +1,7 @@
+# PossibleIncorrectUsageOfComparisonOperator
+
+**Severity Level: Information**
```
I was wondering because we can determine pretty accurately know that the usage of the comparison operator inside if statements is unintentional, if we should up the level to warning because some people ignore/disable warnings of informational level?

LaurentDardenne commented:
About the renaming of the rule,I think there are two different cases.
To separate these two cases, even if it shares the same code of analysis, would allow a granularity in the configuration.

This causes parsing errors:
```Powershell
if ($a > = $b){ "ok"} else { "nok"}
```
And this is valid:
```Powershell
if ($a >=){ "ok"} else { "nok"}
```
This rule can also be applied to 'While' and 'Do' statements.

For the level of the rule, it depends on the level of knowledge of the person.
In my opinion, this writing is reserved for people who know the associated behavior and implicit conversions.
I wonder if this rule will be of any use to them, since here this writing is intentional.
In this case the level 'Warning' will be considered too strict and the level 'Information' would be considered too verbose and useless, the rule will just confirm his intention :-).

For a beginner the level 'Warning' would be useful and the level 'Information' would be too risky if it configures the triggering of the rules on 'Warning'.

And it is easier with PSSA to exlude a rule than to configure its level of criticality.


["Since assignment inside if statements are very rare,"](https://github.com/PowerShell/PSScriptAnalyzer/blob/d432e00c280e2c9e8dfad2466f3fd21afc6e51f9/RuleDocumentation/PossibleIncorrectUsageOfAssignmentOperator.md)
Yes, but it depends on [who](https://github.com/PowerShell/PSScriptAnalyzer/issues/809) :-)

Test with a module triggering 38 times this rule:
```Powershell 
$r=Invoke-ScriptAnalyzer -path 'C:\Program Files\WindowsPowerShell\Modules\InvokeBuild\5.1.0'|
    where-object rulename -eq 'PSPossibleIncorrectUsageOfAssignmentOperator'
#------
$_ = ${*}.Data['Checkpoint.Import']
$r = ${*checkpoint}.Result
$file = $b['File']
$log = $b['Log']
$up = $PSCmdlet.SessionState.PSVariable.Get('*')
$f = [System.IO.Directory]::GetFiles($Path, '*.build.ps1')
$_ = $f -match '[\\/]\.build\.ps1$'
$_ = $env:InvokeBuildGetFile
$_ = $PSCmdlet.SessionState.PSVariable.Get('*')
$_ = $PSBoundParameters['Result']
$_ = $p.Name
$_ = ${*}.All[$Name]
$_ = ${*}.All[$Task]
$_ = $PSCmdlet.GetVariableValue(${*n})
$d = $Hash[$f]
$c = $d.C[$n]
$_ = $PSBoundParameters['Variable']
$$ = $PSCmdlet.GetVariableValue($_)
$_ = $PSBoundParameters['Environment']
$_ = $PSBoundParameters['Property']
$t = ${*}.All[$r]
$t = ${*}.All[$r]
$r = ${*}.All[$_]
${private:*i} = $Task.Inputs
${private:*x} = $Task.If
${*}.Q = $BuildTask -eq '?' -or $BuildTask -eq '??'
${**} = ${*}.All
$_ = $_.Error
$w = ${*}.Warnings
$_ = ${*}.P
$root = ${env:ProgramFiles(x86)}
$r = $paths -like '*\Enterprise\*'
$r = $paths -like '*\Professional\*'
$r = $paths -like '*\Community\*'
$r = $r.parameters
$r = $r.parameter
$d = $Hash[$f]
$c = $d.C[$n]
```

JamesWTruher commented:
while `if ( $a >=) { "ok" }` is parsable, it's probably not what you want, because a file named `=` will be created with the contents of `$a`

bergmeister commented:
@LaurentDardenne 

>About the renaming of the rule,I think there are two different cases.
To separate these two cases, even if it shares the same code of analysis, would allow a granularity in the configuration.

The rule is for cases where the user simply forgets the right PowerShell syntax (due to context switching or because he/she is a newbie) and therefore I do not understand why someone would enable/disable only one of them if the rule were to be split into 2 rules.

Parsing errors are OK and by design (they come from the PowerShell parses itself) and are a good signal to raise an early red flag because if a script does not parse then there is no point running the script at all and the error messages of the parses are good as well and even point to the right line where the problem happens. There is no point to catch parsing errors and do our own logic, we would be re-inventing the wheel.

>> Yes, but it depends on who :-)

> "Since assignment inside if statements are very rare,"

The implementation of the rule is clever enough to not warn in scenarios where assignment is intentional.

> This rule can also be applied to 'While' and 'Do' statements.

That's a good point. We should consider implementing the rule for those statements as well.

Your 38 examples do not trigger the rule for me (I even tried putting some of the lines in an if statement), but they do trigger `PSUseDeclaredVarsMoreThanAssignments` correctly, was this a mistake or can you elaborate please? Examples of false positives that repro would be very welcome.

LaurentDardenne commented:
@bergmeister 
_Can you please give an example/scenario where a split would be needed?_
It's more of a personal choice or a question of granularity. 
In this [discussion](https://github.com/PowerShell/PSScriptAnalyzer/issues/809#issuecomment-321700101) nightroman prefers this syntax :
```powershell
if ($x = <something>) {
```   
When I deliver code to a customer, I prefer it for maintenance / knowledge reasons :
```powershell
$x = <something>
if ($x) {
```   
For that I would like to have this rule with a level 'warning'.
People using the first, more concise syntax will prefer, I suppose, this rule with an 'information' level.

For this case :
```powershell
if ($a > $b){ "ok"} else { "nok"}
``` 
I assume that the majority of people will consider this case as being of the warning level.

This is what makes me think that splitting these cases into two rules,the both  with a  'warning' level , with could satisfy as many people as possible :
```powershell
Invoke-ScriptAnalyzer -ExcludeRule 'PSPossibleIncorrectUsageOfAssignmentOperator' -IncludeRule 'PSPossibleIncorrectUsageOfRedirectionOperator'
Invoke-ScriptAnalyzer -IncludeRule 'PSPossibleIncorrectUsageOfAssignmentOperator','PSPossibleIncorrectUsageOfRedirectionOperator'
#or
Invoke-ScriptAnalyzer #default 
```

_There is no point to catch parsing errors and do our own logic, we would be re-inventing the wheel._
We are talking about the same thing, but not in the same way :-)
```Powershell
Invoke-ScriptAnalyzer -ScriptDefinition 'if ($a > = $b){ "ok"} else { "nok"}'
# Invoke-ScriptAnalyzer : Parse error in script definition:  Unexpected token '$b' in expression or statement at line 1
# column 12.
# At line:1 char:1
# + Invoke-ScriptAnalyzer -ScriptDefinition 'if ($a > = $b){ "ok"} else { ...
# + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#     + CategoryInfo          : ParserError: (UnexpectedToken:String) [Invoke-ScriptAnalyzer], ParseException
#     + FullyQualifiedErrorId : Parse error in script definition:  Unexpected token '$b' in expression or statement at l
#    ine 1 column 12.,Microsoft.Windows.PowerShell.ScriptAnalyzer.Commands.InvokeScriptAnalyzerCommand
```
What I wanted to say : it is not necessary to put this case in the tests because PSSA is waiting for a syntactically correct script.

_Your 38 examples do not trigger the rule for me_
I would check again with a development version posted in Appveyor. so we will be sure to use the same code.
```powershell
ipmo PSScriptAnalyzer
$r=Invoke-ScriptAnalyzer -path 'C:\Program Files\WindowsPowerShell\Modules\InvokeBuild\5.1.0'|
    where rulename -eq 'PSPossibleIncorrectUsageOfAssignmentOperator'
$r.count
#38
$r[0]|select *
# 
# Line                 : 59
# Column               : 7
# Message              : Did you really mean to make an assignment inside an if statement? If you rather meant to check
#                        for equality, use the '-eq'  operator.
# Extent               : $_ = ${*}.Data['Checkpoint.Import']
# RuleName             : PSPossibleIncorrectUsageOfAssignmentOperator
# Severity             : Information
# ScriptName           : Build-Checkpoint.ps1
# ScriptPath           : C:\Program Files\WindowsPowerShell\Modules\InvokeBuild\5.1.0\Build-Checkpoint.ps1
# RuleSuppressionID    :
# SuggestedCorrections :
```

bergmeister commented:
@LaurentDardenne  I think you are missing a detail: The legitimate  case `if ($gotSomething = Get-MeSomething){ }` will not trigger the rule and neither will `if ($x){}`. Only cases like `if ($a=$b){}` or `if ($a>$b){}` will, which is the reason why I am thinking of always treating it as a warning because the likelihood that it is an error is extremely high due and this is why I don't see the need for the split because the rule does not trigger false positives.
I do not have anything like `if ($a > = $b)` in the pester tests that I added because as we both agree, it throws a parser error. You must be referring to the test cases that I wrote in the corresponding issue, you can ignore them now because I wrote them only down to have a big picture of cases where a user might make a mistake.
About your 38 test cases. I build it locally using `.\buildCoreClr.ps1 -Framework net451 -Configuration Release -Build; .\buildCoreClr.ps1 -Framework net451 -Configuration PSV3Release -Build; .\buildCoreClr.ps1 -Framework netstandard1.6 -Configuration Release -Build; .\build.ps1 -BuildDocs` and then I import the psd1 in the out folder (using Windows PowerShell 5.1 or PowerShell Core) and it does not repro on my dev machine.
How are you building/using the PSSA version of my branch? I personally cannot use the artifacts from AppVeyor because I get this error when importing the psd1:
````powershell
Import-Module : Could not load file or assembly 'file:///C:\Users\cberg\Downloads\psscriptanalyzer\out\PSScriptAnalyzer
\Microsoft.Windows.PowerShell.ScriptAnalyzer.dll' or one of its dependencies. Operation is not supported. (Exception
from HRESULT: 0x80131515)
At C:\Users\cberg\Downloads\psscriptanalyzer\out\PSScriptAnalyzer\PSScriptAnalyzer.psm1:24 char:17
+ $binaryModule = Import-Module -Name $binaryModulePath -PassThru
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Import-Module], FileLoadException
    + FullyQualifiedErrorId : System.IO.FileLoadException,Microsoft.PowerShell.Commands.ImportModuleCommand
````

LaurentDardenne commented:
_... this is why I don't see the need for the split..._
Ok.

_How are you building/using the PSSA version of my branch?_ 
I did not use the latest commited version.
Now, I use the zip file [here](https://github.com/bergmeister/PSScriptAnalyzer/tree/4078e41333719d9a7a3a7e50aff3801cb1abe54d)
I compile it with VS studio 2017( VM in Azure) or with the build script.
```powershell
cd C:\temp\pssa\out\PSScriptAnalyzer 
ipmo .\PSScriptAnalyzer.psd1
Invoke-ScriptAnalyzer -ScriptDefinition 'if ($root = ${env:ProgramFiles(x86)}) {}'
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfCompariso Information             1     Did you really mean to use the assignment operator
# nOperator                                                         '=' inside an if statement? If you rather meant to
#                                                                   check for equality, use the '-eq'  operator.
# PSUseDeclaredVarsMoreThanAssignment Warning                 1     The variable 'root' is assigned but never used.
# s

Invoke-ScriptAnalyzer -ScriptDefinition 'if ($root = get-nothing) {}'

# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSUseDeclaredVarsMoreThanAssignment Warning                 1     The variable 'root' is assigned but never used.
# s

Invoke-ScriptAnalyzer -ScriptDefinition 'if ($root > $b) {}'
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfCompariso Warning                 1     Did you really mean to use the redirection operator
# nOperator                                                         '>'? If you wanted to use an equality comparer then
#                                                                   use the '-gt' (greater than) or '-ge' (greater or
#                                                                   equal) operators.


Invoke-ScriptAnalyzer -ScriptDefinition 'if ($root > get-nothing) {}'
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfCompariso Warning                 1     Did you really mean to use the redirection operator
# nOperator                                                         '>'?   ...

Invoke-ScriptAnalyzer -ScriptDefinition 'if (get-nothing > $root) {}'
# 
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfAssignmen Warning                 1     Did you really mean to use the redirection operator
# tOperator                                                          '>'?   ...
```

bergmeister commented:
@LaurentDardenne Thanks for clarifying the repro, very helpful, I can repro now and will take a look at them.

LaurentDardenne commented:
One last remark on the presence of two rules.
In case I want to filter one of the cases for this code :
```Powershell 
$sb={
 if ($root > $b) {}
 if ($root = ${env:ProgramFiles(x86)}) {}
}
```
I have to filter this way :
```Powershell
Invoke-ScriptAnalyzer -ScriptDefinition "$sb"|
 Where {($_.RuleName -eq 'PSPossibleIncorrectUsageOfComparisonOperator') -and ($_.Severity -ne 'Information')}
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfCompariso Warning                 2     Did you really mean to use the redirection operator
# nOperator                                                         '>'? If you wanted to use an equality comparer then
```
Until now this was enough :
```Powershell 
Invoke-ScriptAnalyzer -ScriptDefinition "$sb" -ExcludeRule 'PSPossibleIncorrectUsageOf_AssignmentOperator'
```
Finally the deletion of the application of the rule, for one of the two cases, by the attribute [SuppressMessage](https://github.com/PowerShell/PSScriptAnalyzer/blob/54c4bf765931f668a8b5dad52f0a4455e2a3b84e/README.md#suppressing-rules) is not possible:
```powershell
$sb={
function Test{
[System.Diagnostics.CodeAnalysis.SuppressMessage('PSPossibleIncorrectUsageOfComparisonOperator', '')]
[CmdletBinding()]
 param ()
 if ($root > $b) {}
 if ($root = ${env:ProgramFiles(x86)}) {}
 }
}
Invoke-ScriptAnalyzer -ScriptDefinition "$sb"
#nothing
```

bergmeister commented:
@LaurentDardenne I think I slowly start to understand your point of view and can see now how a split into 2 rules would be beneficial given the possibility of false positives for the assignment case.
I also went through all the samples and I think it is by design that cases of the form `if ($a = $variableExperssionOfSomeForm){}` trigger the rule, there is no way to be more clever without understanding the surrounding script code outside the if statement (except maybe trying to see if the variable is being used inside the statement block of the if statement as an approval that the assignment was intentional). I will definitely look at the following cases of the form `if ($a = InvokeExpressionOfsomeForm)){}` and will try to see if I can avoid triggering the rule in those special cases (similar to how I already did it for `if ($a = Get-MeSomething){ }`):
- `if ($_a= $PSCmdlet.SessionState.PSVariable.Get('*')){ }` and similar
- `if (${*}.Q = $BuildTask -eq '?' -or $BuildTask -eq '??'){ }`
- `if ($r = $paths -like '*\Enterprise\*'){ }`
I think if the redirection operator is used inside an if statement then this must be by accident (since the variable on the left hand side will never get assigned). Can you give an example usage where using the `>` operator is intentional AND useful?
Thanks for being patient.

LaurentDardenne commented:
_I also went through all your example samples_
They come from the InvokeBuilder module, we can also consult the chapter 7.11 Assignment operators of language [specifications](https://www.microsoft.com/en-us/download/details.aspx?id=36389) .

_Can you give an example usage where using the > operator is intentional AND useful?_
No sorry, I never use this write option in an if statement.
I do not know the behavior of such a writing (case of lock on the file or of insufficient rights, with $ErrorActionPreference = 'SilentlyContinue', ...).

But maybe this for the assignment can occur :
```powershell
 $Resgen="$psScriptRoot\ResGen.exe" 
 if ($ResultExec=.$Resgen $Source $Destination 2>&1) {}
```
It's not written like that in the original code and I wonder if this writing has an interest ...

_What do you think about this approach of avoiding false positives?_
Sounds good to me. That said, I think the problem here is not in the depth of analysis but the behavior expected by users. 
If for 
```powershell
if ( $files = get-childitem )
```
@JamesWTruher [consider](https://github.com/PowerShell/PSScriptAnalyzer/pull/859#pullrequestreview-92344281) that _"There is a pattern which is very useful which will cause warnings to appear"_
others consider that this:
```powershell
if ( $A = $B )
```
is also a very useful pattern (see InvokeBuilder module) and [others](https://github.com/PowerShell/EditorSyntax/issues/35#issuecomment-322077137) that it will always be an error. 

Consider a trigger level configuration could cover these cases :

- No trigger : everything is allowed, we respect the possibilities of the Powershell language. Is identical to the -ExcludeRule parameter.
- Trigger for variable assignment only : _if ($ A = $ B)_. Are allowed _if ($ files = get-childitem)_ and derived expressions.
- Trigger always : the mere presence of **'='** into an If statement triggers the rule. 

So we are no longer concerned about the intention it is the user who declares it.
Obviously it would be a global behavior and we could not change the behavior for a portion of code or for the entirety of a file (all except).

bergmeister commented:
To avoid false positives for assignments I refactored the rule to check if variable on LHS is used in statement block and if not, then warn. This should then not leave any false positive cases, therefore there is no need for a rule split any more and the level was bumped to be a warning. @LaurentDardenne Let me know if this works well for you now as well.

LaurentDardenne commented:
@bergmeister  
The contain of the _Extend.Text_ property does not appear correctly, we lose the contents of the statement that triggers the error.
The _PossibleIncorrectUsageOfComparisonOperator.md_ file should contain an explicit example :
 ($ a = $ b -> $ a -eq $ b)

bergmeister commented:
The highlighting of the `Extent.Text` to be as close as possible to the operator is by design because this is where people should see the squigglies in VsCode.
````powershell
$w=Invoke-ScriptAnalyzer -ScriptDefinition 'if($a > $b){}'
$w.Extent.Text
> $b
$w=Invoke-ScriptAnalyzer -ScriptDefinition 'if($a > $b){}'
$w.Extent.Text
=
$w=Invoke-ScriptAnalyzer -ScriptDefinition 'if($a == $b){}'
$w.Extent.Text
==
````

JamesWTruher commented:
@bergmeister @LaurentDardenne I'm not sure whether we've reached consensus here - I'm looking for a clear signal on this before taking it. Eventually, we'll have a pragma way to exclude a rule from applicability. If we take this now, there may be a small group which will need to exclude this rule. Is that ok?

bergmeister commented:
To summarise from my side (I have also updated the PR summary):
- The rule to warn against usage of `=` inside if statements will now only trigger if the variable being assigned on the LHS is not being used later in the statement block.
- I do not foresee any false positive about the warning when using `>` inside if statements unless a counter example is given
Maybe it is better to enable this new rule not by default? Or try using pre-releases to get early feedback?
I am happy to take any criticism

JamesWTruher commented:
@bergmeister - we've got merge conflicts now - could you rebase and push?

bergmeister commented:
@imfrancisd has added new ideas in his comments [here](https://github.com/PowerShell/PSScriptAnalyzer/issues/809#issuecomment-367430931) and [here](https://github.com/PowerShell/PSScriptAnalyzer/issues/809#issuecomment-367552359) about rather using additional parenthesis as a mechanism to be used as a way of stating that the assignment is intentional. It is good to have new and different ideas but I am not sure how the community would react to this especially because it means that people would need to adapt their scripts first for this implicit rule suppression. Therefore I referenced his comments in here to get other people's opinions.

LaurentDardenne commented:
I think the idea of _imfrancisd_ is to be considered.
I prefer to use 2 parentheses instead of placing [attributes](https://github.com/PowerShell/Plaster/blob/master/build.settings.ps1#L12).

Into Windows Powershell specification ,  7.1.1 Grouping parentheses
```
To write to the pipeline the value of any expression containing top-level side effects, enclose that expression in parentheses, as follows:
($a = 1234 * 3.5)		# pipeline gets 4319
As such, the grouping parentheses in this case are not redundant.
```
The IF statement does not require double parentheses.

imfrancisd commented:
### First Things First

@LaurentDardenne Thanks! I like the parentheses more than attributes too. Other language communities that have this problem (assignment in if condition) have adopted this coding convention and have formalized it in their analyzers, so I thought PSScriptAnalyzer and the PowerShell community might consider adopting that convention too.

@bergmeister I apologize for not posting my comments here first. I did not know about this pull request and even after I found out about it, it seemed weird that the person who asked for this feature and the person who first showed concerns about this feature were not in the discussion. At least as far as I know.

### Ok, now that that's out of the way

My problem with the approach of checking the use of the variable to determine = vs. -eq is that it makes this script okay:

```
if ($x = $y) {
    do-something -with $x
}
```

However, that is exactly the kind of script that this rule is supposed to flag.

https://github.com/PowerShell/PSScriptAnalyzer/issues/809#issue-249487289
@jongross4 requested this feature with

> When moving from other languages where a simple '=' equal sign can be used both as assignment and operator, a warning should be logged. Other Boolean operations should be flagged too. Examples:
> if ($x = $y) {} -- Always true
> for ($i = 1; $i > 99; $i++) - alias to out-file

He did not write the body of "if ($x = $y) {}", but in real scripts I bet $x is used inside that if block more often than not.

bergmeister commented:
I posted a twitter poll [here](https://twitter.com/ChristophBergm/status/966777114413404160) to involve more people in the community to comment on whether we rather want to have rather more warnings with possible false positives and catch more cases or only call out cases where we are more sure that it is unintentional.

bergmeister commented:
The community has spoken:
![image](https://user-images.githubusercontent.com/9250262/36633909-c713a1e6-1994-11e8-97f5-0d9b4f372dbc.png)
I posted and discussed this also in the PowerShell slack forums (the global and the UK one) and it seems that the additional option of being able to indicate deliberate assignment via extra parenthesis is preferred (most people said that they do not like to litter their code with suppressions and at the moment PSSA does not even support suppression for just one line/statement). Therefore I will attempt to implement this.
From the discussions and poll, the outcome was also to rather ditch the approach of not warning when the LHS variables is referenced in some way. I will revert back to what we originally had but will try to make it not flag when using special syntax that represents statements like `[System.IO]::GetTempPath` similar to how I already included usage of cmdlets like `Get-Foo` on the RHS. No exclusions will however apply if the RHS is a variable or variable expression.

bergmeister commented:
@JamesWTruher  @LaurentDardenne @imfrancisd 
I added:
-More exceptions for PossibleIncorrectUsageOfAssignmentOperator to also exclude usage of  `[NameSpace]::Method()` or `$b -match $c` on the RHS of the if expression. The rule should continue to warn if any kind of variable expression is on the RHS.
-Add optional, implicit clang suppression when extra parenthesis are used as requested by the community
-Add additional PossibleIncorrectUsageOfRedirectionOperator rule to have 2 separate rules since this rule does not have any false positives.
Therefore the PR is again up for review

imfrancisd commented on Tests/Rules/PossibleIncorrectUsageOfAssignmentOperator.tests.ps1:
```diff
@@ -40,23 +40,38 @@ Describe "PossibleIncorrectUsageOfAssignmentOperator" {
-        It "returns no violations when there is an evaluation on the RHS" {
-            $warnings = Invoke-ScriptAnalyzer -ScriptDefinition 'if ($a = Get-ChildItem){}' | Where-Object {$_.RuleName -eq $ruleName}
+        It "returns no violations when using implicit clang style suppresion" {
+            $warnings = Invoke-ScriptAnalyzer -ScriptDefinition 'if ( ($a -eq $b) ){ }' | Where-Object {$_.RuleName -eq $ruleName}
```
Should be = instead of -eq to check clang style suppression
```
if ( ($a = $b) ){}
```

imfrancisd commented:
@bergmeister 
> -More exceptions for PossibleIncorrectUsageOfAssignmentOperator to also exclude usage of  [NameSpace]::Method() or $b -match $c on the RHS of the if expression. The rule should continue to warn if any kind of variable expression is on the RHS.

Is the exception of -match because it sometimes returns a boolean and sometimes it returns an array? If so, what about operators that are purely arithmetic and operators that are purely boolean?

```
    #Will this be flagged?
    #Common in loops with index.
    #
    if ($i = $array.Count -1) {}

    #Will this be flagged?
    #Common in functions with multiple switches to configure behavior.
    #
    if ($switch1 = ($switch2 -and $switch3)) {}
```

I think those lines should be flagged.

bergmeister commented:
@imfrancisd The current exceptions to `PossibleIncorrectUsageOfAssignmentOperator` are if one of the following asts are present on the right hand side:
- `CommandAst` (a powershell cmdlet/function)
- `InvokeMemberExpressionAst` to include other function calls (.net methods or methods on object)
- `BinaryExpressionAst` (operators, like -match, -and, `-`)
So yes, your 2 examples would not get flagged due to them being a `BinaryExpressionAst`. The idea was that anything that is any sort of function or operation/transformation on data gets excluded because it could be by design. But I am happy to remove `BinaryExpressionAst` from the exclusion list because this case is much less likely to be an assignment by design. Thanks for the feedback. :-)

LaurentDardenne commented:
I do not find this use case in the test file :
 ```powershell
Invoke-ScriptAnalyzer -ScriptDefinition 'if ( ($a=$b) ){}'
```

bergmeister commented:
Thanks @LaurentDardenne for spotting that, I fixed it.  It seems I was unlucky to introduce a regression both in the implementation and test when doing the final refactoring.

LaurentDardenne commented:
@bergmeister
 _Only those who do nothing make no mistakes_ ;-)

LaurentDardenne commented:
Should not the second example fail, even if we use () ?
```powershell
Invoke-ScriptAnalyzer -ScriptDefinition 'if ($_ == $_.Error) {}'
# RuleName                            Severity     ScriptName Line  Message
# --------                            --------     ---------- ----  -------
# PSPossibleIncorrectUsageOfAssignmen Warning                 1     Did you mean to use the assignment operator '='? The
# tOperator                                                         equals operator in PowerShell is 'eq'.

Invoke-ScriptAnalyzer -ScriptDefinition 'if (($_ == $_.Error)) {}'
#no warning
```
typo in the error message :
```
...The equals operator in PowerShell is '-eq'.
```

bergmeister commented:
@LaurentDardenne The clang style extra parenthesis are a way of letting the author state implicitly that assignment is deliberate and by design for`PSPossibleIncorrectUsageOfAssignmentOperator`without "having to litter the code with suppressions"(that was what people said in the Slack forums, which is understandable, especially since we cannot suppress one line/statement only at the moment).
I changed the warning text to rather use the term `equality operator`, I just used the other expression because to me it is the equal/equals operator because that's what the `-eq` stands for.

imfrancisd commented:
@bergmeister What do you think about "if ($null = )"?

Shouldn't they always be flagged?
```
Command : invoke-scriptanalyzer -script 'if ($null = $b) {}'
Output  : 
          RuleName                            Severity     ScriptName Line  Message                                                     
          --------                            --------     ---------- ----  -------                                                     
          PSPossibleIncorrectUsageOfAssignmen Information             1     Did you mean to use the assignment operator '='? The        
          tOperator                                                         equality operator in PowerShell is 'eq'.                    
          
          
          

Command : invoke-scriptanalyzer -script 'if ($null = [string]::Empty) {}'
Output  : 
          RuleName                            Severity     ScriptName Line  Message                                                     
          --------                            --------     ---------- ----  -------                                                     
          PSPossibleIncorrectUsageOfAssignmen Information             1     Did you mean to use the assignment operator '='? The        
          tOperator                                                         equality operator in PowerShell is 'eq'.                    
          
          
          

Command : invoke-scriptanalyzer -script 'if ($null = [string]::Concat($a, $b)) {}'
Output  : 

Command : invoke-scriptanalyzer -script 'if ($null = $hashtable.Count) {}'
Output  : 
          RuleName                            Severity     ScriptName Line  Message                                                     
          --------                            --------     ---------- ----  -------                                                     
          PSPossibleIncorrectUsageOfAssignmen Information             1     Did you mean to use the assignment operator '='? The        
          tOperator                                                         equality operator in PowerShell is 'eq'.                    
          
          
          

Command : invoke-scriptanalyzer -script 'if ($null = $hashtable["key"]) {}'
Output  : 
          RuleName                            Severity     ScriptName Line  Message                                                     
          --------                            --------     ---------- ----  -------                                                     
          PSPossibleIncorrectUsageOfAssignmen Information             1     Did you mean to use the assignment operator '='? The        
          tOperator                                                         equality operator in PowerShell is 'eq'.                    
          
          
          

Command : invoke-scriptanalyzer -script 'if ($null = $hashtable.Item("key")) {}'
Output  : 

Command : invoke-scriptanalyzer -script 'if ($null = get-childitem) {}'
Output  : 

Command : invoke-scriptanalyzer -script 'if ($null = (get-childitem)) {}'
Output  : 

Command : invoke-scriptanalyzer -script 'if ($null = $(get-childitem)) {}'
Output  : 

Command : invoke-scriptanalyzer -script 'if ($null = @(get-childitem)) {}'
Output  : 
```

bergmeister commented:
@imfrancisd The rule currently does not inspect the LHS of the assignment, only the RHS. But it would be a good suggested improvement because a common (recommended pattern) is to do comparison via `$null -eq $foo` and using $null on the LHS does not make sense inside an assignment, therefore I will consider this improvement.

imfrancisd commented:
@bergmeister 
> Does it look ready to you now since I would like to finish this PR soon.

I didn't know I had a say in the matter! :-)

Not my place to say. I just wander around GitHub issues from time to time and one day found myself here in pull request land.

bergmeister commented:
@imfrancisd I rather meant ready in terms of whether you are happy with the cases where it warns and where it does not. Everyone in the community can have a say that we will take into account and this PR shows that we can come up with a solution where everyone is more happy. The final decision and technical review will still be by James but we are always open to feedback since we do this for the community (not even Microsoft makes any money with PSScriptAnalyzer)

bergmeister commented:
@JamesWTruher After community feedback I addressed it and implemented additional enhancements for special cases. To me the rule looks very polished now. Can you please re-review and give your opinion?

bergmeister commented:
@JamesWTruher  After community feedback I addressed it and implemented additional enhancements for special cases. To me the rule looks very polished now. Can you please re-review and give your opinion?

bergmeister commented on RuleDocumentation/PossibleIncorrectUsageOAssignmentOperator.md:
`````diff
@@ -0,0 +1,54 @@
+````
+
+```` PowerShell
+if ($a = Get-Something) # Only execute action if command returns something and assign result to variable
`````
I know what you mean but it was a strong desire of the community to not see warnings when assignment is by design as this is a common use case. See referenced issue and there was also a discussion on the testing channel in Slack. I can sort of understand this fear/annoyance of people about false positives, especially given that one cannot suppress on a per-line basis in PSSA. I guess it is better to warn in most cases and help save the developer time rather than always warning and annoying the developer.

bergmeister commented on RuleDocumentation/PossibleIncorrectUsageOfRedirectionOperator.md:
`````diff
@@ -0,0 +1,29 @@
+### Correct
+
+```` PowerShell
+if ($a -eq $b)
`````
Yes, this makes more sense to someone reading the document and comparing wrong vs right. I will change that.

bergmeister commented on Rules/ClangSuppresion.cs:
```diff
@@ -0,0 +1,24 @@
+    internal static class ClangSuppresion
+    {
+        /// <summary>
+        /// The community requested an implicit suppression mechanism that follows the clang style where warnings are not issued if the expression is wrapped in extra parenthesis
```
OK.

bergmeister commented on Rules/PossibleIncorrectUsageOfRedirectionOperator.cs:
```diff
@@ -0,0 +1,99 @@
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// PossibleIncorrectUsageOfRedirectionOperator: Warn if someone uses '>' or '>=' inside an if or elseif statement because in most cases that is not the intention.
```
We actually do, but the comment got out of date. I will fix that. I am following the current comment style but my feeling is that the codebase has too many repeated comments.
